internal/neofs/tree.go

```diff
@@ -209,6 +317,98 @@ func (c *TreeClient) Close() error {
 	return nil
 }
 
+func (c *TreeClient) addVersion(ctx context.Context, cnrID *cid.ID, treeID, attrPath, filepath string, version *layer.NodeVersion) error {
+	path := strings.Split(filepath, separator)
+	meta := map[string]string{
+		oidKV:    version.OID.String(),
+		attrPath: path[len(path)-1],
+	}
+
+	if version.IsDeleteMarker {
+		meta[isDeleteMarkerKV] = "true"
+	}
+
+	if version.IsUnversioned {
+		meta[isUnversionedKV] = "true"
+
+		node, err := c.getUnversioned(ctx, cnrID, treeID, filepath)
+		if err == nil {
+			parentID, err := c.getParent(ctx, cnrID, treeID, node.ID)
+			if err != nil {
+				return err
+			}
+
+			return c.moveNode(ctx, cnrID, treeID, version.ID, parentID, meta)
+		}
+
+		if !errors.Is(err, layer.ErrNodeNotFound) {
+			return err
+		}
+	}
+
+	return c.addNodeByPath(ctx, cnrID, treeID, path[:len(path)-1], meta)
+}
+
+func (c *TreeClient) getVersions(ctx context.Context, cnrID *cid.ID, treeID, filepath string, onlyUnversioned bool) ([]*layer.NodeVersion, error) {
+	keysToReturn := []string{oidKV, isUnversionedKV, isDeleteMarkerKV}
+	path := strings.Split(filepath, separator)
+	nodes, err := c.getNodes(ctx, cnrID, treeID, fileNameKV, path, keysToReturn, false)
+	if err != nil {
+		if strings.Contains(err.Error(), "not found") {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("couldn't get nodes: %w", err)
+	}
+
+	result := make([]*layer.NodeVersion, 0, len(nodes))
+	for _, node := range nodes {
+		nodeVersion, err := newNodeVersion(node)
+		if err != nil {
+			return nil, err
+		}
+
+		if onlyUnversioned && !nodeVersion.IsUnversioned {
+			continue
+		}
+
+		result = append(result, nodeVersion)
+	}
+
+	return result, nil
+}
+
+func (c *TreeClient) getParent(ctx context.Context, cnrID *cid.ID, treeID string, id uint64) (uint64, error) {
+	request := &tree.GetSubTreeRequest{
+		Body: &tree.GetSubTreeRequest_Body{
+			ContainerId: cnrID[:],
+			TreeId:      treeID,
+			RootId:      id,
+			BearerToken: getBearer(ctx),
+		},
+	}
+
+	if err := c.signRequest(request.Body, func(key, sign []byte) {
+		request.Signature = &tree.Signature{
+			Key:  key,
+			Sign: sign,
+		}
+	}); err != nil {
+		return 0, err
+	}
+
+	cli, err := c.service.GetSubTree(ctx, request)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get sub tree client: %w", err)
+	}
+
+	resp, err := cli.Recv()
+	if err != nil {
+		return 0, fmt.Errorf("failed to get sub tree: %w", err)
+	}
+
+	return resp.GetBody().GetParentId(), nil
```

Comment:
I don't think it is good idea to abandon stream, this can lead to issues on the server side, like this one: https://github.com/nspcc-dev/neofs-node/issues/1270. Let's read it all and ignore all other other `resp` messages if we don't need them.